File: 백준/Silver II/11724. 연결 요소의 개수/연결 요소의 개수.py

import sys
sys.setrecursionlimit(10**6)

def operator(i, visited, graph):
    visited[i] = True

    for node in graph[i]:
        if visited[node] == False:
            visited = operator(node, visited, graph)
    return visited

def solution(n, graph):
    visited = [False] * (n+1)
    answer = 0

    for i in range(1, n+1):
        if visited[i] == False:
            operator(i, visited, graph)
            answer += 1

    return answer

n, m = map(int, sys.stdin.readline().split())

graph = [[] for _ in range(n+1)]

for i in range(m):
    node1, node2 = map(int, sys.stdin.readline().split())
    graph[node1].append(node2)
    graph[node2].append(node1)


print(solution(n, graph))

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.util.*;
import java.io.*;

public class Main {
    static int n;
    static int m;
    static List<Integer>[] graph;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        graph = new List[n+1];
        for(int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for(int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());

            int node1 = Integer.parseInt(st.nextToken());
            int node2 = Integer.parseInt(st.nextToken());

            graph[node1].add(node2);
            graph[node2].add(node1);
        }

        boolean[] visited = new boolean[n+1];
        int cnt = 0;
        for(int i = 1; i <= n; i++) {
            if(visited[i]) continue;

            dfs(i, visited);
            cnt++;
        }

        System.out.println(cnt);
    }

    private static void dfs(int node, boolean[] visited) {
        visited[node] = true;

        for(Integer next : graph[node]) {
            if(visited[next]) continue;
            dfs(next, visited);
        }
    }
}
